ratis-server-api/src/main/java/org/apache/ratis/server/protocol/RaftServerProtocol.java

```diff
@@ -38,4 +40,6 @@ enum Op {REQUEST_VOTE, APPEND_ENTRIES, INSTALL_SNAPSHOT}
   InstallSnapshotReplyProto installSnapshot(InstallSnapshotRequestProto request) throws IOException;
 
   StartLeaderElectionReplyProto startLeaderElection(StartLeaderElectionRequestProto request) throws IOException;
+
+  ReadIndexReplyProto readIndex(ReadIndexRequestProto request) throws IOException;
```

Comment:
@SzyWilliam , thanks a lot for working on this! Just have the following comment:

Since `LeaderStateImpl.getReadIndex()` is asynchronous, let define the this method in `RaftServerAsynchronousProtocol` instead.  Otherwise, it requires a thread to `waitForReadIndex(..)`.  It will limit the number of concurrent requests and potentially block the server when all the threads are calling `waitForReadIndex(..)`.